File: app/src/main/java/com/lzm/lightLive/http/CustomGsonConvertFactory.java
package com.lzm.lightLive.http;

import static java.nio.charset.StandardCharsets.UTF_8;
import android.util.Log;
import androidx.annotation.NonNull;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okio.Buffer;
import retrofit2.Converter;
import retrofit2.Retrofit;

public class CustomGsonConvertFactory extends Converter.Factory{

    private static final String TAG = "CustomGsonConvertFactory";

    private final Gson gson;

    private final ResponseConverter converter;

    public abstract static class ResponseConverter {
        protected abstract String convertResponse(ResponseBody value) throws IOException;
    }

    private CustomGsonConvertFactory(Gson gson, ResponseConverter converter) {
        if (gson == null) throw new NullPointerException("gson == null");
        this.gson = gson;
        this.converter = converter;
    }

    public static CustomGsonConvertFactory create(ResponseConverter converter) {
        return create(new Gson(), converter);
    }

    public static CustomGsonConvertFactory create(Gson gson, ResponseConverter converter) {
        return new CustomGsonConvertFactory(gson, converter);
    }

    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(@NonNull Type type,
                                                            @NonNull Annotation[] annotations,
                                                            @NonNull Retrofit retrofit) {
        TypeAdapter<?> adapter = gson.getAdapter(TypeToken.get(type));
        return new CustomGsonConvertFactory.CustomGsonResponseBodyConverter<>(gson, adapter, type, converter);
    }

    @Override
    public Converter<?, RequestBody> requestBodyConverter(Type type, @NonNull Annotation[] parameterAnnotations,
                                                          @NonNull Annotation[] methodAnnotations, @NonNull Retrofit retrofit) {
        TypeAdapter<?> adapter = gson.getAdapter(TypeToken.get(type));
        return new CustomGsonConvertFactory.CustomGsonRequestBodyConverter<>(gson, adapter, type);
    }

    static final class CustomGsonRequestBodyConverter<T> implements Converter<T, RequestBody> {
        private final MediaType MEDIA_TYPE = MediaType.parse("application/json; charset=UTF-8");
        private final Charset UTF_8 = StandardCharsets.UTF_8;

        private final Gson gson;
        private final TypeAdapter<T> adapter;
        private final Type type;

        CustomGsonRequestBodyConverter(Gson gson, TypeAdapter<T> adapter, Type type) {
            this.gson = gson;
            this.adapter = adapter;
            this.type = type;
        }

        @Override
        public RequestBody convert(T value) throws IOException {
            Buffer buffer = new Buffer();
            Writer writer = new OutputStreamWriter(buffer.outputStream(), UTF_8);
            JsonWriter jsonWriter = gson.newJsonWriter(writer);
            adapter.write(jsonWriter, value);
            jsonWriter.close();
            return RequestBody.create(MEDIA_TYPE, buffer.readByteString());
        }
    }

    public static class CustomGsonResponseBodyConverter<T> implements Converter<ResponseBody, T> {

        private final Gson gson;
        private final TypeAdapter<T> adapter;
        private final Type type;
        private final ResponseConverter responseConverter;

        public CustomGsonResponseBodyConverter(Gson gson, TypeAdapter<T> adapter,
                                               Type type, ResponseConverter responseConverter) {
            this.gson = gson;
            this.adapter = adapter;
            this.type = type;
            this.responseConverter = responseConverter;
        }

        @Override
        public T convert(@NonNull ResponseBody value) throws IOException {
            String response = responseConverter.convertResponse(value);
            MediaType contentType = value.contentType();
            Charset charset = contentType != null ? contentType.charset(UTF_8) : UTF_8;
            if(null != response) {
                value.close();
            }
            InputStream inputStream = new ByteArrayInputStream(response.getBytes());
            Reader reader = new InputStreamReader(inputStream, charset);
            JsonReader jsonReader = gson.newJsonReader(reader);
            try {
                return adapter.read(jsonReader);
            } finally {
                value.close();
            }
        }
    }
}

File: app/src/main/java/com/lzm/lightLive/http/CustomGsonConvertFactory.kt
package com.lzm.lightLive.http

import com.google.gson.Gson
import com.google.gson.TypeAdapter
import com.google.gson.reflect.TypeToken
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import okio.Buffer
import retrofit2.Converter
import retrofit2.Retrofit
import java.io.*
import java.lang.reflect.Type
import java.nio.charset.StandardCharsets

class CustomGsonConvertFactory private constructor(gson: Gson?, converter: ResponseConverter) :
    Converter.Factory() {
    private val gson: Gson
    private val converter: ResponseConverter

    abstract class ResponseConverter {
        @Throws(IOException::class)
        abstract fun convertResponse(value: ResponseBody): String
    }

    override fun responseBodyConverter(
        type: Type,
        annotations: Array<Annotation>,
        retrofit: Retrofit
    ): Converter<ResponseBody, *> {
        val adapter = gson.getAdapter(TypeToken.get(type))
        return CustomGsonResponseBodyConverter<Any>(gson, adapter, type, converter)
    }

    override fun requestBodyConverter(
        type: Type, parameterAnnotations: Array<Annotation>,
        methodAnnotations: Array<Annotation>, retrofit: Retrofit
    ): Converter<*, RequestBody> {
        val adapter = gson.getAdapter(TypeToken.get(type))
        return CustomGsonRequestBodyConverter<Any>(gson, adapter, type)
    }

    internal class CustomGsonRequestBodyConverter<T>(
        private val gson: Gson,
        private val adapter: TypeAdapter<out T>,
        private val type: Type
    ) : Converter<T, RequestBody> {
        private val mediaType = "application/json; charset=UTF-8".toMediaTypeOrNull()
        private val charset = StandardCharsets.UTF_8
        @Throws(IOException::class)
        override fun convert(value: T): RequestBody {
            val buffer = Buffer()
            val writer: Writer = OutputStreamWriter(buffer.outputStream(), charset)
            val jsonWriter = gson.newJsonWriter(writer)
            adapter.write(jsonWriter, value as Nothing?)
            jsonWriter.close()
            return buffer.readByteString().toRequestBody(mediaType)
        }
    }

    class CustomGsonResponseBodyConverter<T>(
        private val gson: Gson, private val adapter: TypeAdapter<out T>,
        private val type: Type, private val responseConverter: ResponseConverter
    ) : Converter<ResponseBody, T> {
        @Throws(IOException::class)
        override fun convert(value: ResponseBody): T {
            val response = responseConverter.convertResponse(value)
            val contentType = value.contentType()
            val charset =
                if (contentType != null) contentType.charset(StandardCharsets.UTF_8) else StandardCharsets.UTF_8
            value.close()
            val inputStream: InputStream = ByteArrayInputStream(response.toByteArray())
            val reader: Reader = InputStreamReader(inputStream, charset)
            val jsonReader = gson.newJsonReader(reader)
            return value.use {
                adapter.read(jsonReader)
            }
        }
    }

    companion object {
        fun create(converter: ResponseConverter): CustomGsonConvertFactory {
            return create(Gson(), converter)
        }

        fun create(gson: Gson?, converter: ResponseConverter): CustomGsonConvertFactory {
            return CustomGsonConvertFactory(gson, converter)
        }
    }

    init {
        if (gson == null) throw NullPointerException("gson == null")
        this.gson = gson
        this.converter = converter
    }
}